File: 0053-maximum-subarray/0053-maximum-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function(nums) {
    let max = nums[0]
    let curMax = nums[0]
    for (let i=1; i < nums.length; i++) {
        curMax = Math.max(curMax + nums[i], nums[i])
        max = Math.max(max, curMax)
    }
    return max
};

maxSubArray = function(nums) {
    const n = nums.length
    const dp = nums.slice()
    for (let i=1; i < n; i++) {
        dp[i] = Math.max(dp[i] + dp[i-1], dp[i])
    }
    return Math.max(...dp)
}

File: 0053-maximum-subarray/0053-maximum-subarray.kt
class Solution {
    fun maxSubArray(nums: IntArray): Int {
        val n = nums.size
        val dp = IntArray(n) {0}
        var max = nums[0]
        dp[0] = nums[0]
        for (i in 1..n-1) {
            dp[i] = Math.max(nums[i], dp[i-1] + nums[i])
            max = Math.max(max, dp[i])
        }
        return max
    }
}